Jakarta rewrite infrastructure

gsmet opened the pull request:
Hey @manovotn @Sanne @maxandersen @n1hility @radcortez ,

Here is the promised PR with the Jakarta rewrite infrastructure. It required quite a lot of experiments but I'm quite happy with the result (it looks far more simple in the end than all the experiments I had to do to get to this result). The hardest part was to find a way to apply the changes without ending up with something you can't build anymore as OpenRewrite needs the Maven build to work.

I went until the ArC extension in this first PR. My next step is to be able to build the Hibernate Validator one as it shouldn't be too hard on the HV side but will unblock quite a lot of dependencies as a lot of them are built on the same model and will require the same dependencies for them to work.
It should be far faster now that I figured out a way to orchestrate this thing.

I added a `jakarta/README.md` explaining the general approach and I would be happy to present the approach to you anytime.

Note that this requires (but good news, everything is taken care of in the `jakarta/transform.sh` script so you don't have to do anything related to this):
- a patched version of OpenRewrite (I'm contributing my patches back there so that's until they merged them)
- a patched version of the Rewrite Maven Plugin
- a patched version of the Kotlin Maven Plugin (yes, I know... but I need to be able to skip the main compilation which the Kotlin plugin does not support, it only supports `maven.test.skip` for the test compilation)
- the script also builds SmallRye Config while waiting for the CR2 promised by Roberto

Please see the `jakarta/README.md` before experimenting with this as it has a few advices on how to experiment.

Let's wait for CI on this one as I had to do some more build fixes in addition to the ones that I already merged. These build fixes are in separate commits.

Questions welcome. Happy to present this work in a call.

gsmet committed:
Add OpenRewrite infrastructure and rewrite the bootstrap module

gsmet committed:
Remove quarkus-jsonp-deployment dependency from test-framework/common

gsmet committed:
Use a new approach by running OpenRewrite globally

This simplifies things a lot but requires a few adjustments to be able
to skip compilation during the run.

gsmet committed:
Move unpack-dependencies to prepare-package in core/launcher

gsmet committed:
Add documentation for Jakarta rewrite

Sanne commented on jakarta/README.md:
```diff
@@ -0,0 +1,103 @@
+You can for instance use the following approach:
+
+- create a `quarkus-jakarta` copy of your `quarkus` repository
+- add a `jakarta` remote to this `quarkus-jakarta` copy: `git remote add jakarta ../quarkus/`
```
suggestion:
`git worktree add ../quarkus-jakarta`

maxandersen commented on jakarta/README.md:
````diff
@@ -0,0 +1,103 @@
+
+The active recipes have to be declared in the target POM with a `plugin` block in the `<build><plugins>` section similar to:
+
+```xml
````
could we not avoid all the pom.xml changes by putting this plugin config in parent and have it pickup `rewrite.yml` ?

maxandersen replied:
ack - its just that if there is a rewrite.yml next to the pom.xml my jbang based runner for rewrite could be used in a parallel and removed need to change each pom.xml ...just add rewrite.yml where needed.

Sanne commented:
It's failing for me with:
```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.377 s
[INFO] Finished at: 2022-03-28T15:01:17+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] No plugin found for prefix 'rewrite' in the current project and in the plugin groups [org.apache.maven.plugins, org.codehaus.mojo] available from the repositories [local (/home/sanne/.m2/repository), jboss-public-repository (https://repository.jboss.org/nexus/content/groups/public/), central (https://repo.maven.apache.org/maven2)] -> [Help 1]
```

Is that a reasonable clue for you?

Sanne commented:
Attaching the full logs:
[transform.log.gz](https://github.com/quarkusio/quarkus/files/8363648/transform.log.gz)

gsmet commented:
OK, so the problem Sanne has is that they released a new version of the rewrite-maven-plugin since I started working on this and it messes things up.
I forked it too to make things more stable.

gsmet committed:
Set export REWRITE_OFFLINE=true to disable fetching external projects

Obviously, you need to have a previous build complete so that you have
all pieces around.
Also you might miss the latest adjustments to OpenRewrite so make sure
you do a full build from time to time.

Sanne commented:
Excellent, works fine for me as well now.

gsmet committed:
Switch to SmallRye Config 3.0.0-RC2

radcortez commented:
Excellent work @gsmet. Thanks!

gsmet commented:
CI was green. I'm force pushing to squash a !fixup commit.
